### ISSUE TITLE
Comments about Lexer

### ISSUE THREAD

**Issue by KubaO** [thyringer/zuse#1]
The way you did the lexer certainly works, but it's IMHO unreadable. Regexes are at least semi-readable, but state machines - not so much. I would not mess with the prelexer and so on. Scanning should be done by the regexp engine provided by the language. After specifying the lexical structure using regexes, the only code left to write is then stuffing this data into Token objects. A C lexer fits into a single file, and not a very long one (<1000 lines) when done this way. And is eminently readable.

The simplest way to make a lexer is to write a bunch of regexps with named capture groups. For example:

```ts
const ws = "(?<ws>[ \\t\\n\\r\\v\\h]+)"
const comment = "(?<comment>#[^\\n]*)"
const single_quote_string = "'(?:\\\\.|[^'])*'"
const double_quote_string = "\"(?:\\\\.|[^\"])*\""
const string = `(?<string>${single_quote_string}|${double_quote_string})`

const tokens = [ws, comment, string].join("|")

let regex = new RegExp(tokens)
```

The name of the capture group can be converted to the token kind/type and so on when the source is iterated over with the regex.

In post-lex phase, the indent and dedent tokens are synthesized for languages that need those, based on the line and column information of preceding tokens.

In my limited experience, it makes most sense not to deal with whitespace tokens directly in the token sequence fed to the parser. But losing whitespace and comments is not nice either. Instead, each Token carries a `ws` field that has an optional Token for preceding whitespace. You can also have a `comment` field for a post-comment for documentation, etc. That depends on the language - I didn't investigate Kalkyl much.

Below is a trivial lexer in Python. Putting all token definitions into a `TokenKind` enum may not be always feasible, but it's IMHO a decent starting point. You already have done a lot of this, e.g. `PunctuationData` etc. It's very simple to generate regexes out of those, with named capture groups.

This excerpt comes from https://github.com/KubaO/resilient-ll-parsing-py/blob/main/resilient.py

```py
def kw(name: str) -> str:
    return f"{name}(?![_a-zA-Z0-9])"


class TokenKind(Enum):
    # Token types recognized by the regexp-based scanner
    Arrow = '->'
    LParen = r'\('
    RParen = r'\)'
    LCurly = '{'
    RCurly = '}'
    Eq = '='
    Semi = ';'
    Comma = ','
    Colon = ':'
    Plus = r'\+'
    Minus = '-'
    Star = r'\*'
    Slash = '/'
    Int = '[0-9]+'

    FnKeyword = kw('fn')
    LetKeyword = kw('let')
    ReturnKeyword = kw('return')
    TrueKeyword = kw('true')
    FalseKeyword = kw('false')

    Name = '[_a-zA-Z0-9]+'
    Ws = r'[ \\n\\t\\h\\v\\r]+'
    Error = '.'

    Eof = None


class Token(NamedTuple):
    kind: TokenKind
    text: str


TYPES = tuple(t for t in TokenKind if isinstance(t.value, str))
SCANNER = re.compile("|".join(f"(?P<{t.name}>{t.value})" for t in TYPES))


def scan(text: str) -> Iterable[Token]:
    error = ""

    def error_token() -> Token:
        nonlocal error
        result, error = Token(TokenKind.Error, error), ""
        return result

    for match in SCANNER.finditer(text):
        assert match  # the scanner always matches, with a fallback of Error
        i = match.lastindex - 1
        text = match[i + 1]
        match TYPES[i]:
            case TokenKind.Error:
                error += text
            case TokenKind.Ws:
                pass  # drop whitespace
            case ty:
                if error: yield error_token()
                yield Token(ty, text)
    if error: yield error_token()
```

**Comment by KubaO** [comment]
Making the parser resilient so that it's usable in a language server for an IDE will be another issue. Alex Kladov's post abut resilient LL parsing covers the basics: https://matklad.github.io/2023/05/21/resilient-ll-parsing-tutorial.html

**Comment by thyringer** [comment]
Thank you for your detailed feedback. I have now merged the "prelexer" with "lexer" and added the missing recording of special lexemes. The reason for this separation was the considerable complexity. But I managed to fix this quite well now, so that I only have a simple loop that tokenizes line by line in one go.

The states are simply a necessity in order to know whether the lexer is currently reading a string, word, etc., or just spaces. They are basically just a summary of similar lexeme categories. I could of course use the lexeme categories such as "integer" or "raw string" directly instead of these general states, but then I would have significantly more handlers and code duplications, because then the respective characters such as space, tab, numbers, letters, punctuation, etc. also have to be handled for each category.

The approach with regular expressions is of course feasible, but I wanted a clean approach instead for a simple loop that reads in character by character and remembers the context in order to efficiently capture the tokens.

My lexer now only has about 800 lines of code, + 400 lines that describe the token type and store data, e.g. which punctuation characters are "delimiting", i.e. can directly border other lexemes, such as brackets. I managed to express these complex things using a simple map that the lexer can access. So if I were to add a new operator in the future, I would only have to add it to the map in the "Token.ts" file and specify two boolean values ​​​​how this operator behaves in relation to other lexemes, without having to touch the lexer itself.

To be fair, my language is much bigger than C or Go: more operators, keywords and various literals, e.g. there are binary and hexadecimal data strings, or decimal numbers with repeating ends (in parentheses directly following numbers). At the same time, my lexer already carries out all checks and returns errors, whereby an invalid lexeme is simply scanned further and only noted whether it is an incorrect token in order to then ignore it in the later translation process so that further analyses are still possible. So yes, there is already resilience to enable a later language server.

Finally, it should be said that my goal was to have a solid lexer that can be easily expanded later and rewritten in Kalkyl. Therefore, I also separated the lexer data from the actual logic implemented by independent functions, instead of stuffing everything into the lexer class and working with tons of "this".

**Comment by KubaO** [comment]
That makes sense. Perhaps then, since regexes are so useful, you could
implement a very basic regex engine and use that for lexing. It really
makes life super-easy, the notation is compact, and you'd be
probably having other uses for regexes. All the regexes can go into a text
file for example.

On the other hand - I'm not super sure that putting keywords in the lexer
is the best way to go. It puts the same information into two places: in the
lexer and then in the parser. It is the parser that cares about what a
particular identifier is. I would just make sure that the keywords can all
be captured under the IDENTIFIER category or something like that. Then the
parser will be expecting a particular keyword and that can be given as a
string for simplicity. After all, keywords have localized uses in the
parser. The vast majority of them will appear only once. So going through
the brouhaha of having special tokens for those provides IMHO little
benefit.

Keep working on the project - it is very nice! I'm working on a relay
computer that is not terribly dissimilar from Zuse's Z4. A hundred words of
RAM, and a much larger operand stack - Z4 only had up to two operands on
the stack. It would be cool for me to take your compiler and retarget it to
the relay computer :)

**Comment by thyringer** [comment]
Yeah, I've found that it doesn't make much sense to isolate the individual steps strictly, because the data structures are closely related to each other. For example, I have the maps that describe keywords and punctuation, in a separate file, and develop the lexer and parser in the same directory (I had originally planned a more extensive project structure).

However, I don't want to put any more work into the lexer now, because it's finished, easily expandable, and serves its purpose (and I've gained enough experience with it), but I want to develop the parser matching my language so I can start generating code straight away. Things like static type systems, code checks, and other analyses will be added step by step, based on practical experience. So my goal is to quickly get a presentable prototype of the compiler.
—
And thank you very much!

Your project sounds very nice too. But I have to admit, I don't know much about hardware; I would therefore never program a compiler backend myself that spits out any specific assembler, but just use C as the target language; and if that works one day, maybe WASM for web stuff (provided that this offers better performance than the detour through C) and LLVM bitcode as an alternative (but without integrating LLVM / simply directly as compiler output in files that can be further translated with clang).

**Comment by KubaO** [comment]
No problem. I will tackle the backend. Have done it a few times. And I agree that it’s best to leave that lexer alone. It works no was more spilling my thoughts than anything. It’s a cool project and I’m glad you posted it!20. des. 2024 kl. 7:25 pm skrev ER ***@***.***>:﻿
Yeah, I've found that it doesn't make much sense to isolate the individual steps strictly, because the data structures are closely related to each other. For example, I have the maps that describe keywords and punctuation, in a separate file, and develop the lexer and parser in the same directory (I had originally planned a more extensive project structure).
However, I don't want to put any more work into the lexer now, because it's finished, easily expandable, and serves its purpose (and I've gained enough experience with it), but I want to develop the parser matching my language so I can start generating code straight away. Things like static type systems, code checks, and other analyses will be added step by step, based on practical experience. So my goal is to quickly get a presentable prototype of the compiler.
—
And thank you very much!
Your project sounds very nice too. But I have to admit, I don't know much about hardware; I would therefore never program a compiler backend myself that spits out any specific assembler, but just use C as the target language; and if that works one day, maybe WASM for web stuff (provided that this offers better performance than the detour through C) and LLVM bitcode as an alternative (but without integrating LLVM / simply directly as compiler output in files that can be further translated with clang).
